### ISSUE TITLE
Tick inside setTimeout callback results in negative time-travel

### ISSUE THREAD

**Issue by thw0rted** [jasmine/jasmine#1929]
## Current Behavior
If you call `tick` from inside a Clock-managed (fake) timeout, it can advance the `MockDate` closure-scoped `currentTime` past the goal time for the outer `tick` call, then [*subtract* from `currentTime` as it returns up the call chain](https://github.com/jasmine/jasmine/blob/main/src/core/DelayedFunctionScheduler.js#L171).  The net effect is one where code can get a value from `new Date().getTime()` that's larger than a later call, that is, traveling back in time.

## Expected Behavior
Use of the Jasmine clock / MockDate should never violate the fundamental assumption that `new Date().getTime()` monotonically increases.

## Possible Solution
Never subtract from `currentTime`? (This probably happens for a reason, though.)  If it helps, I think it's possible that this was introduced in trying to fix #915, maybe as interplay with the fix for #1190, but that could be a red herring.

## Suite that reproduces the behavior (for bugs)
```ts
// Code under test
function runWork(work: () => boolean) {
    while (work()) {
        console.log("Did work", new Date().getTime());
        // ... check for early exit
    }
}

describe("runWork", () => {
    let clock: jasmine.Clock;
    let lastWork = 0;

    // Make a worker that takes a little time and tracks when it finished
    function mockWork(times: number): () => boolean {
        return () => {
            clock.tick(1);
            const now = new Date().getTime();
            expect(lastWork).toBeLessThan(now, "New work should always be later");
            lastWork = now;
            times--;
            return times > 0;
        }
    }

    beforeEach(() => {
        clock = jasmine.clock();
        clock.install();
        clock.mockDate(new Date(1));
    });

    it("breaks the space-time continuum", () => {
        const work1 = mockWork(3);
        setTimeout(() => {
            runWork(work1);
        }, 1);
        clock.tick(1);
        expect(lastWork).toBeLessThanOrEqual(new Date().getTime(), "Past happened after present?!?");

        const work2 = mockWork(3);
        setTimeout(() => {
            runWork(work2);
        }, 1);
        clock.tick(1);
        expect(lastWork).toBeLessThanOrEqual(new Date().getTime(), "Past happened after present again?!?");
    });
});
```

## Context
I have code that (very reasonably) assumes that `old = new Date().getTime(); {do some stuff}; diff = new Date().getTime() - old;` can never produce a negative value for `diff`, and this bug breaks that assumption.

## Your Environment
* Version used: 3.6.4
* Environment name and version (e.g. Chrome 39, node.js 5.4): node 14.17.3
* Operating System and version (desktop or mobile): WIn10 latest

**Comment by sgravrock** [comment]
I think the basic issue here is that `DelayedFunctionScheduler#tick` (and thus `Clock#tick`) is not reentrant:

```
    self.tick = function(millis, tickDate) {
      millis = millis || 0;
      var endTime = currentTime + millis;
      var oldCurrentTime = currentTime;

      runScheduledFunctions(endTime, tickDate);
      currentTime = endTime;  // <--- potential negative time travel
    };

```

If any of the scheduled functions calls `tick`, the inner call will set `currentTime` to the later time and then the outer call will set it to the earlier one, thus causing the backwards time travel. The `runScheduledFunctions` function that does the bulk of the work in `DelayedFunctionScheduler` also assumes that nothing else changes the current time while it's running, and could misbehave in interesting ways if that assumption doesn't hold.

I can think of a few ways to patch around that, but ultimately I'm not sure it makes sense to change the current time while partway through changing the current time. How should that behave? I'm not sure what the effect of advancing the clock by 1ms while in the middle of advancing it by 500ms should be. The best thing to do might be to issue an error in that case rather than trying to define how it behaves.

I suggest looking for a way to move the calls to `tick()` up to the top level of the spec so that they're not inside a `setTimeout` or similar.

**Comment by thw0rted** [comment]
This makes sense and if the answer is "don't call `tick` inside a timeout", that's fair, but should be enforced by the framework.  (I think it's going to make my mocks harder to reason about, but in the past that's just meant that I made bad mocks, or bad data flow, which, as a root cause, wouldn't surprise me at all.)

What I expected it to do was, have one consistent "current simulated time" that `tick` always advances.  After any call to `tick(x)`, I expect the next statement to happen at least `x` milliseconds after the previous one (thought not necessarily exactly `x`).

Suppose the spec says

````ts
setTimeout(a, 10);
setTimeout(b, 12);
tick(10);
console.log(new Date().getTime());
````

If the timeout handler `a` calls `tick(5)`, I expect the log line to happen at simulated time 15, after `b` runs.  I *think* this is how it would work if you changed your snippet to

```js
      runScheduledFunctions(endTime, tickDate);
      if (endTime > currentTime) {
        currentTime = endTime;
      }
```

Basically, if a scheduled function advances simulation time past the "goal" of the outer `tick` call, there's no need to do anything.  Does that break anything else?

**Comment by sgravrock** [comment]
> This makes sense and if the answer is "don't call tick inside a timeout", that's fair, but should be enforced by the framework. (I think it's going to make my mocks harder to reason about, but in the past that's just meant that I made bad mocks, or bad data flow, which, as a root cause, wouldn't surprise me at all.)

I agree in the abstract, and if I was going to build this subsystem from scratch today I'd certainly make ticking inside a timeout be an error. The problem with making it an error now is that there's an un-knowable number of specs out there that rely on the current behavior. I hear that you're not happy about having to change your specs to not tick inside of a timeout, and I don't blame you. Now imagine that you don't understand this issue, but you have to make the same kind of change in a bunch of old crusty legacy specs you've never looked at before because the latest version of Jasmine started giving you an error.

I have no way of knowing how many how many people rely on the existing behavior, but I'd guess that it's a lot. Ticking inside of a timeout is a fairly intuitive thing to do, and 5 years is enough time to accumulate a lot of specs that do it.

> If the timeout handler a calls tick(5), I expect the log line to happen at simulated time 15, after b runs. 

What about this?

```
setTimeout(() => jasmine.clock().tick(5), 7);
jasmine.clock().tick(10);
console.log(new Date().getTime());
```

Does that output 10, 12, or 15?

> Basically, if a scheduled function advances simulation time past the "goal" of the outer tick call, there's no need to do anything. Does that break anything else?

I'm not sure if it breaks anything else, but it looks like it doesn't fix the problem. I applied the below patch to the current head of main and the example specs still fail.

```
diff --git a/lib/jasmine-core/jasmine.js b/lib/jasmine-core/jasmine.js
index ab9bb00e..dd8b753d 100644
--- a/lib/jasmine-core/jasmine.js
+++ b/lib/jasmine-core/jasmine.js
@@ -3633,7 +3633,9 @@ getJasmineRequireObj().DelayedFunctionScheduler = function(j$) {
       var endTime = currentTime + millis;
 
       runScheduledFunctions(endTime, tickDate);
-      currentTime = endTime;
+      if (endTime > currentTime) {
+        currentTime = endTime;
+      }
     };
 
     self.scheduleFunction = function(
diff --git a/spec/core/DelayedFunctionSchedulerSpec.js b/spec/core/DelayedFunctionSchedulerSpec.js
index d06a66c1..fb13c021 100644
--- a/spec/core/DelayedFunctionSchedulerSpec.js
+++ b/spec/core/DelayedFunctionSchedulerSpec.js
@@ -279,3 +279,55 @@ describe('DelayedFunctionScheduler', function() {
     expect(tickDate).toHaveBeenCalledWith(1);
   });
 });
+
+describe('Example suite that reproduces the bug', function() {
+  function runWork(work) {
+    while (work()) {
+      console.log("Did work", new Date().getTime());
+      // ... check for early exit
+    }
+  }
+
+  describe("runWork", () => {
+    let clock;
+    let lastWork = 0;
+
+    // Make a worker that takes a little time and tracks when it finished
+    function mockWork(times) {
+      return () => {
+        clock.tick(1);
+        const now = new Date().getTime();
+        expect(lastWork).toBeLessThan(now, "New work should always be later");
+        lastWork = now;
+        times--;
+        return times > 0;
+      }
+    }
+
+    beforeEach(() => {
+      clock = jasmine.clock();
+      clock.install();
+      clock.mockDate(new Date(1));
+    });
+
+    afterEach(function() {
+      jasmine.clock().uninstall();
+    });
+
+    it("breaks the space-time continuum", () => {
+      const work1 = mockWork(3);
+      setTimeout(() => {
+        runWork(work1);
+      }, 1);
+      clock.tick(1);
+      expect(lastWork).toBeLessThanOrEqual(new Date().getTime(), "Past happened after present?!?");
+
+      const work2 = mockWork(3);
+      setTimeout(() => {
+        runWork(work2);
+      }, 1);
+      clock.tick(1);
+      expect(lastWork).toBeLessThanOrEqual(new Date().getTime(), "Past happened after present again?!?");
+    });
+  });
+})
diff --git a/src/core/DelayedFunctionScheduler.js b/src/core/DelayedFunctionScheduler.js
index eb84c48a..8ae277ee 100644
--- a/src/core/DelayedFunctionScheduler.js
+++ b/src/core/DelayedFunctionScheduler.js
@@ -12,7 +12,9 @@ getJasmineRequireObj().DelayedFunctionScheduler = function(j$) {
       var endTime = currentTime + millis;
 
       runScheduledFunctions(endTime, tickDate);
-      currentTime = endTime;
+      if (endTime > currentTime) {
+        currentTime = endTime;
+      }
     };
 
     self.scheduleFunction = function(
```

**Comment by thw0rted** [comment]
Re: changing current behavior, I totally understand.  Is there a precedent for at least issuing a (non-failing) warning instead?  Like, the first time a spec calls `tick` during a delayed function, at least log something?  I don't know if you have a "rule" against warnings like that e.g. because some people have specs that examine console output.

Re: your other example, it should output 12:

```js
setTimeout(() => {
  console.log("b", new Date().getTime()); // b7
  jasmine.clock().tick(5);
  console.log("c", new Date().getTime()); // c12
}, 7);
console.log("a", new Date().getTime()); // a0
jasmine.clock().tick(10);
console.log("d", new Date().getTime()); // d12
```

So, you'd get `a0 / b7 / c12 / d12`. No time travel, easy (?) to reason about.

I haven't had a chance to try fixing this myself, but I'll take your word for it that my idea didn't help.  If I could fix it, would you accept a PR?  It would obviously be a breaking change -- mind you, breaking in the interest of the integrity of space-time 😁.  Maybe it could hide behind a flag passed to `tick`, or an option passed to `jasmine.clock()`, or just a property on the Clock instance (`jasmine.clock().neverTimeTravel = true`)?

**Comment by sgravrock** [comment]
> Re: changing current behavior, I totally understand. Is there a precedent for at least issuing a (non-failing) warning instead? Like, the first time a spec calls tick during a delayed function, at least log something?

Yes. We already send deprecation warnings to `console.warn`, and I think it makes sense to do that here too.

> I haven't had a chance to try fixing this myself, but I'll take your word for it that my idea didn't help. If I could fix it, would you accept a PR? It would obviously be a breaking change -- mind you, breaking in the interest of the integrity of space-time 😁. Maybe it could hide behind a flag passed to tick, or an option passed to jasmine.clock(), or just a property on the Clock instance (jasmine.clock().neverTimeTravel = true)?

In principle, yes. If I understand correctly we're talking about making the clock value be correct when certain timers fire, and potentially firing more timers as a consequence of fixing the clock. That seems like it should be fine for 4.0. I can't really commit to that yet, though, because it's so hard to evaluate the impact of changes in this area without having working code to play with. But if you're willing to run the risk of getting a no after you've already put in the work, I'm willing to work with you to try to make it a yes.

I'd prefer to avoid putting this behind a flag (the mock clock is already complicated enough) but that could be an option if it looks like this will have impact that's difficult for users to deal with.

**Comment by sgravrock** [comment]
@thw0rted I'm not sure if you're still working on this or what approach you're taking, but FYI I'm planning to get 4.0 out the door in early January, possibly as early as Jan 1. After that, any solution to this would need to not be a breaking change.

**Comment by thw0rted** [comment]
Honestly, this fell off my plate after we last discussed it and I haven't circled back around.  I have some other loose ends to tie up today but might be able to take a look this evening or tomorrow.  I'll at least re-read this history to see if I can figure out what I'd like to see happen, and try to get together "red" tests that describe sane/consistent behavior for these cases soon.

**Comment by thw0rted** [comment]
Sorry to push this so close to your deadline, but I think I may have cracked it.  My initial approach was right, I just didn't grok the entirety of the DelayFunctionScheduler.

In practice, `currentTime` is updated every time `tickDate` is called, it just wasn't obvious because the calls weren't actually adjacent to each other.  I've fixed that in [my fork here](https://github.com/thw0rted/jasmine/tree/4.0) -- it was basically the change I posited but carried over each of the 3 cases where `tickDate` is called.  I also added the test you used, and it now passes, without breaking any other tests.

**Comment by thw0rted** [comment]
PR is linked, it's pretty small so maybe you can still look it over before 4.0 launches?

I did realize something while writing the PR description: `Clock#tick` does not enforce that the tick amount is non-negative, nor do the docs say that it has to be.  There are no tests covering negative ticking, but that doesn't mean some yahoo didn't build a suite that relies on it working the way it currently does.  (My take is of course "don't do that, silly" but it's not my project, so...)

**Comment by sgravrock** [comment]
Yes, I can take a look at the PR before 4.0.

I'm not worried about breaking negative tick amounts. Just as a gut call, that seems obviously incorrect and I'm having a hard time coming up with a scenario where someone would do that on purpose. (I also don't have any idea what the existing implementation does with negative tick amounts. Probably nothing good.)
